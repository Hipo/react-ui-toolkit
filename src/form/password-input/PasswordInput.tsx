import "./_password-input.scss";

import React, {useState} from "react";
import classNames from "classnames";

import Input from "../input/Input";
import {InputProps} from "../input/util/inputTypes";
import Button from "../../button/Button";

export interface PasswordInputProps extends Omit<InputProps, "leftIcon" | "rightIcon"> {
  hideIcon?: React.ReactNode;
  revealIcon?: React.ReactNode;
}

function PasswordInput({
  testid,
  name,
  value,
  onChange,
  hasError,
  placeholder,
  customClassName,
  hideIcon,
  revealIcon
}: PasswordInputProps) {
  const [isPasswordShown, setPasswordVisibility] = useState(false);
  const passwordInputIconClassName = classNames("password-input__icon", {
    "password-input__icon--is-visible": Boolean(value)
  });
  const passwordInputClassName = classNames("password-input", customClassName);
  let iconAriaLabel = "Show password";
  let icon = revealIcon;
  let inputType: "password" | "text" = "password";

  if (isPasswordShown) {
    iconAriaLabel = "Hide password";
    icon = hideIcon;
    inputType = "text";
  }

  return (
    <Input
      testid={testid}
      type={inputType}
      name={name}
      customClassName={passwordInputClassName}
      value={value}
      hasError={hasError}
      placeholder={placeholder}
      onChange={onChange}
      rightIcon={
        revealIcon &&
        hideIcon && (
          <Button
            testid={`${testid}-password-visibility-icon`}
            customClassName={passwordInputIconClassName}
            aria-label={iconAriaLabel}
            onClick={togglePasswordVisibility}
            shouldStopPropagation={false}
            shouldPreventDefault={false}
          >
            {icon}
          </Button>
        )
      }
    />
  );

  function togglePasswordVisibility() {
    setPasswordVisibility(!isPasswordShown);
  }
}

export default PasswordInput;
