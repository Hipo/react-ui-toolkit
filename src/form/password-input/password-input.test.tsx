import React from "react";
import {render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

import PasswordInput, {PasswordInputProps} from "../password-input/PasswordInput";
import {testA11y} from "../../core/utils/test/testUtils";

describe("<PasswordInput />", () => {
  const defaultPasswordInputProps: PasswordInputProps = {
    testid: "password-input",
    name: "password-input",
    onChange: jest.fn(),
    placeholder: "Test",
    hideIcon: (
      <svg data-testid={"password-input.hide"} height={"16"} width={"16"}>
        <circle
          cx={"50"}
          cy={"50"}
          r={"40"}
          stroke={"black"}
          strokeWidth={"3"}
          fill={"red"}
        />
      </svg>
    ),
    revealIcon: (
      <svg data-testid={"password-input.reveal"} height={"16"} width={"16"}>
        <circle
          cx={"50"}
          cy={"50"}
          r={"40"}
          stroke={"black"}
          strokeWidth={"3"}
          fill={"green"}
        />
      </svg>
    )
  };

  it("should render correctly", () => {
    render(<PasswordInput {...defaultPasswordInputProps} />);
  });

  it("should pass a11y test", async () => {
    const {container} = render(<PasswordInput {...defaultPasswordInputProps} />);

    await testA11y(container);
  });

  it("should hide password initially", () => {
    const {container} = render(<PasswordInput {...defaultPasswordInputProps} />);

    const iconButton = screen.getByRole("button");

    expect(screen.getByPlaceholderText("Test")).toHaveAttribute("type", "password");
    expect(container).toContainElement(iconButton);
    expect(iconButton).toHaveAttribute("aria-label", "Show password");
  });

  it("should toggle password visibility on click icon", () => {
    const {container} = render(<PasswordInput {...defaultPasswordInputProps} />);

    const iconButton = screen.getByRole("button");

    userEvent.click(iconButton);

    expect(screen.getByPlaceholderText("Test")).toHaveAttribute("type", "text");
    expect(container).toContainElement(iconButton);
    expect(iconButton).toHaveAttribute("aria-label", "Hide password");

    userEvent.click(iconButton);

    expect(screen.getByPlaceholderText("Test")).toHaveAttribute("type", "password");
    expect(container).toContainElement(iconButton);
    expect(iconButton).toHaveAttribute("aria-label", "Show password");
  });
});
