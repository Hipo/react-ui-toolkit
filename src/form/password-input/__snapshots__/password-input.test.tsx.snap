// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordInput /> should match snapshot 1`] = `
<div
  className="input-container password-input input-container--type-password"
  data-testid="password-input"
>
  <input
    autoComplete="off"
    autoCorrect="off"
    className="input"
    name="password-input"
    onChange={[Function]}
    placeholder="Test"
    type="password"
  />
  <span
    className="input-container__icon input-container__right-icon"
  >
    <button
      aria-label="Show password"
      className="button password-input__icon"
      data-testid="password-input-password-visibility-icon"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        data-testid="password-input.reveal"
        height="16"
        width="16"
      >
        <circle
          cx="50"
          cy="50"
          fill="green"
          r="40"
          stroke="black"
          strokeWidth="3"
        />
      </svg>
    </button>
  </span>
</div>
`;
